### PR TITLE
Improve `AttributesMixin`

### PR DESCRIPTION
**Pull request by SergeyTsimfer**

### CODE REVIEW

File: seismiqb/src/labels/horizon_attributes.py [hunk 1]
```diff
@@ -40,20 +64,127 @@ class AttributesMixin:
...
+        return array
+
+    def matrix_set_ndim(self, matrix, ndim):
+        """ !!. """
+        if ndim == matrix.ndim:
+            pass
+        elif ndim == matrix.ndim + 1:
+            matrix = matrix[..., np.newaxis]
+        else:
+            msg = f'Result ndim is unequal to requested ndim: `{matrix.ndim} != {ndim}`.'
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:139]: uninformative error message

previous was better

`
            msg = f"Result ndim is {array.ndim}, while requested ndim is {res_ndim}. "\
                  f"Adding more than one new axis is not currently implemented."
`

File: seismiqb/src/labels/horizon_attributes.py [hunk 2]
```diff
@@ -343,30 +359,42 @@ def get_array_values(self, array, shifts=None, grid_info=None, width=5, axes=(2,
...
+        """ Alias for cube values. Depending on what loaded to cube geometries
+        might actually not be amplitudes, so use it with caution.
+        """
+        return self.cube_values
+
+    @property
+    def cube_values(self):
+        """ Values from the cube along the horizon. """
+        cube_values = self.get_cube_values(window=1)
+        cube_values[~self.presence_matrix] = np.nan
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:374]: already done in `get_cube_values`

File: seismiqb/src/labels/horizon_attributes.py [hunk 3]
```diff
@@ -343,30 +359,42 @@ def get_array_values(self, array, shifts=None, grid_info=None, width=5, axes=(2,
...
+    def horizon_metrics(self):
+        """ Calculate :class:`~HorizonMetrics` on demand. """
+        # pylint: disable=import-outside-toplevel
+        from ..metrics import HorizonMetrics
+        return HorizonMetrics(self)
+
+    @property
+    def instantaneous_phase(self):
+        """ Phase along the horizon. """
+        return self.get_instantaneous_phases()[:, :, 0]
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:387]: squeeze

**Review by cdtn — COMMENTED** [review]
Make a tutorial notebook with every attribute visualized

File: seismiqb/src/labels/horizon_attributes.py [hunk 4]
```diff
@@ -10,50 +11,220 @@
...
+        If provided, then used at points where the horizon is absent.
+    dtype : numpy dtype, optional
+        If provided, then dtype of a matrix is changed, and fill_value at absent points is changed accordingly.
+    normalize : bool, str, optional
+        If `min-max` or True, then use min-max scaling.
+        If `mean-std`, then use mean-std scaling.
+        If False, don't scale matrix.
+    n_components : number, optional
+        If integer, then number of components to keep for PCA transformation.
+        If float in (0, 1) range, then amount of variance to be explained after PCA transformation.
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:41]: Poorly implemented

File: seismiqb/src/labels/horizon_attributes.py [hunk 5]
```diff
@@ -420,44 +473,44 @@ def load_attribute(self, src, location=None, **kwargs):
...
     def _load_attribute(self, src, location=None, use_cache=True, **kwargs):
         """ Load horizon attribute at requested location. """
-        try:
-            method_name = self.ATTRIBUTE_TO_METHOD[src]
-        except KeyError as e:
-            raise ValueError(f"Unknown `src` {src}. Expected one of {list(self.ATTRIBUTE_TO_METHOD.keys())}.") from e
+        src = self.ALIAS_TO_ATTRIBUTE.get(src, src)
+
+        if src in self.ATTRIBUTE_TO_METHOD:
+            print('loaded by get_*')
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:486]:
```suggestion
```

File: seismiqb/src/labels/horizon_attributes.py [hunk 6]
```diff
@@ -420,44 +473,44 @@ def load_attribute(self, src, location=None, **kwargs):
...
+        """ !!. """
+        data = getattr(self, src, None)
+        if data is None:
+            aliases = list(self.ALIAS_TO_ATTRIBUTE.keys())
+            raise ValueError(f'Unknown `src` {src}. Expected a matrix-property or one of {aliases}.')
+        return data
 
     @lru_cache(maxsize=1, apply_by_default=False, copy_on_return=True)
+    @transformable
     def get_instantaneous_amplitudes(self, window=23, depths=None, **kwargs):
```
Comment (cdtn) [inline seismiqb/src/labels/horizon_attributes.py:514]: `depths` is bad naming, do smth about it

File: seismiqb/src/labels/horizon.py [hunk 7]
```diff
@@ -799,6 +799,18 @@ def add_to_mask(self, mask, locations=None, width=3, alpha=1, **kwargs):
         return mask
 
 
+    def load_slide(self, loc, axis=0, width=3):
+        """ Create a mask at desired location along supplied axis. """
+        axis = self.geometry.parse_axis(axis)
+        locations = self.geometry.make_slide_locations(loc, axis=axis)
+        shape = np.array([(slc.stop - slc.start) for slc in locations])
+        width = width or max(5, shape[-1] // 100)
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply seismiqb/src/labels/horizon.py:807]: I don't wanna

File: seismiqb/src/labels/horizon_attributes.py (hunk 6 above)
Comment (SergeyTsimfer) [reply seismiqb/src/labels/horizon_attributes.py:514]: No u

**Review by cdtn — APPROVED** [review]

**Review by cdtn — APPROVED** [review]

**Pull request merged**